synapse/storage/databases/main/events_worker.py

```diff
@@ -736,35 +736,80 @@ async def get_stripped_room_state_from_event_context(
             for e in state_to_include.values()
         ]
 
+    def _start_fetch_thread(self) -> None:
+        """Starts an event fetcher."""
+        with self._event_fetch_lock:
+            if (
+                self._event_fetch_list
+                and self._event_fetch_ongoing < EVENT_QUEUE_THREADS
+            ):
+                self._event_fetch_ongoing += 1
+                event_fetch_ongoing_gauge.set(self._event_fetch_ongoing)
+                should_start = True
+            else:
+                should_start = False
+
+        async def _do_fetch() -> None:
+            """Services requests for events from the `_event_fetch_list` queue."""
+            exc = None
+            try:
+                await self.db_pool.runWithConnection(self._do_fetch)
+            except Exception as e:
+                exc = e
+                raise
+            finally:
+                should_restart = False
+                failed_event_list = []
+                with self._event_fetch_lock:
+                    self._event_fetch_ongoing -= 1
+                    event_fetch_ongoing_gauge.set(self._event_fetch_ongoing)
+
+                    if self._event_fetch_ongoing == 0 and self._event_fetch_list:
```

Comment:
similarly, if there *is* an exception, it looks like we leave the remaining entries in the queue for other threads to pick up - but they may not do so for up to `EVENT_QUEUE_TIMEOUT_S`. That's not an implausible thing to do (backing off after an exception sounds good), but let's comment that it's happening.

Reply:
> if there is no exception, is it correct that we only restart if `_event_fetch_ongoing == 0` ?
Yes and no
No, because if there are other threads remaining, `_event_fetch_list` only gets processed once one of those threads completes its current work, however long that takes.
Yes, because otherwise we could restart spuriously if `_enqueue_events` has already started a worker for the new work. This can happen because `_maybe_start_fetch_thread` doesn't claim the work when it starts the thread.

Reply:
I've rejigged it to always restart on a clean exit when there is more work remaining and added a bunch of comments.